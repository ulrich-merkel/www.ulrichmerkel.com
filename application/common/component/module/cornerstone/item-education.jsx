/* eslint-disable immutable/no-mutation, react/jsx-indent */
/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @requires react
 * @requires prop-types
 * @requires classnames
 * @requires shortid
 * @requires common/component/element/headline
 * @requires common/component/element/paragraph
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import shortid from 'shortid';

import Headline from '../../element/headline';
import P from '../../element/paragraph';

/**
 * Function representing a component to return a single react child element.
 *
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
function ModuleCornerstoneItemEducation(props) {
    const {
        cssModifier,
        offset,
        headline,
        lead,
        timeStart,
        timeEnd,
        description,
        place,
        ...otherProps
    } = props;

    const composedListItemClassName = classnames(
        cssModifier
            ? `m-cornerstone__item--${cssModifier}`
            : 'm-cornerstone__item',
        offset ? `has-offset--${offset}` : ''
    );

    return (
        <li
            className={composedListItemClassName}
            itemProp="itemListElement"
            itemScope
            itemType="https://schema.org/EducationEvent"
            {...otherProps}
        >
            <div className="m-cornerstone__description">
                <div className="m-cornerstone__description-content">
                    <Headline
                        className="m-cornerstone__headline"
                        itemProp="name"
                        htmlElement="h4"
                    >
                        {headline}
                    </Headline>
                    <P
                        className="m-cornerstone__company"
                        itemProp="alternateName"
                    >
                        <strong>{lead}</strong>
                    </P>
                    <P className="m-cornerstone__time">
                        (
                        <time className="c-time" itemProp="startDate">
                            {timeStart}
                        </time>{' '}
                        -{' '}
                        <time className="c-time" itemProp="endDate">
                            {timeEnd}
                        </time>
                        )
                    </P>
                    {description.map((text) => {
                        /**
                         * DangerouslySetInnerHTML due to reacts double escaping. Otherwise html elements
                         * are not possible to be set via translation strings.
                         *
                         * @see {@link https://facebook.github.io/react/tips/dangerously-set-inner-html.html}
                         */
                        return (
                            <P
                                key={shortid.generate()}
                                className="m-cornerstone__text"
                                itemProp="description"
                                dangerouslySetInnerHTML={{ __html: text }}
                            />
                        );
                    })}
                </div>
            </div>
            <div className="m-cornerstone__bubble" />
            <div
                hidden
                itemProp="location"
                itemScope
                itemType="https://schema.org/Place"
            >
                <meta itemProp="name" content={place.name} />
                <div
                    itemProp="address"
                    itemScope
                    itemType="http://schema.org/PostalAddress"
                >
                    <meta
                        itemProp="streetAddress"
                        content={place.streetAddress}
                    />
                    <meta
                        itemProp="addressLocality"
                        content={place.addressLocality}
                    />
                    <meta itemProp="sameAs" content={place.sameAs} />
                </div>
            </div>
        </li>
    );
}

/**
 * Valiate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 * @property {string} [headline] - The item headline
 * @property {string} [lead] - The item subline
 * @property {string} [timeStart] - The item start time
 * @property {string} [timeEnd] - The item end time
 * @property {Array} [description=[]] - The items description
 * @property {object} [place={}] - The items place description
 * @property {string} [cssModifier] - The bem css modifier
 * @property {string} [offset] - The css top offset to display items nice
 */
ModuleCornerstoneItemEducation.propTypes = {
    headline: PropTypes.string, // eslint-disable-line react/require-default-props
    lead: PropTypes.string, // eslint-disable-line react/require-default-props
    timeStart: PropTypes.string, // eslint-disable-line react/require-default-props
    timeEnd: PropTypes.string, // eslint-disable-line react/require-default-props
    description: PropTypes.array, // eslint-disable-line react/forbid-prop-types
    place: PropTypes.objectOf(
        PropTypes.oneOfType([PropTypes.string, PropTypes.object])
    ),
    cssModifier: PropTypes.string, // eslint-disable-line react/require-default-props
    offset: PropTypes.string // eslint-disable-line react/require-default-props
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 * @see ModuleCornerstoneItemEducation.propTypes
 */
ModuleCornerstoneItemEducation.defaultProps = {
    description: [],
    place: {}
};

export default ModuleCornerstoneItemEducation;
