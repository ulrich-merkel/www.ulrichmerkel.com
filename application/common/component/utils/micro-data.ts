/**
 * Es6 module for creating mirco data attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type ItemTypeAttributes = {
    itemScope: boolean;
    itemType: string;
};

/**
 * Get component properties for itemProp and itemType.
 *
 * @param {boolean} itemType - The itemType property if available
 * @returns {object|null} The component props for spreading
 */
export function getItemTypeAttributes(
    itemType?: string
): ItemTypeAttributes | null {
    return isValidString(itemType)
        ? {
              itemScope: true,
              itemType
          }
        : null;
}
