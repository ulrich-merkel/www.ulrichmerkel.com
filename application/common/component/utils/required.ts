/**
 * Es6 module for creating form input required attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isBoolean } from 'lodash';

type RequiredAttributes = {
    'aria-required': boolean;
    required: boolean;
};

/**
 * Get component properties for required form input elements.
 *
 * @param {boolean} required - Whether the input element is required or not
 * @returns {object|null} The component props for spreading
 */
export function getRequiredAttributes(
    required?: boolean
): RequiredAttributes | null {
    return isBoolean(required) && required
        ? {
              'aria-required': true,
              required: true
          }
        : null;
}
