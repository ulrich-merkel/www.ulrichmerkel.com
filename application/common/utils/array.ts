/**
 * Es6 module for handling array related utilities.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */

/**
 * Check if given "input" is a valid array with at least one entry.
 *
 * @param {Array<*>} array - The array to be checked
 * @returns {boolean} Whether this is an array with some entries
 */
export function isValidArray(array?: any[]): boolean {
    return Array.isArray(array) && array.length > 0;
}
