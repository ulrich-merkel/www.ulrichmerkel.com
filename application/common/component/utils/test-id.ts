/**
 * Es6 module for creating test id related attributes
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type TestIdAttributes = {
    'data-testid': string;
};

/**
 * Get component properties for testing.
 *
 * @param {string} testId - The test id to be added
 * @returns {object|null} The component props for spreading
 */
export function getTestIdAttributes(testId?: string): TestIdAttributes | null {
    return isValidString(testId)
        ? {
              'data-testid': testId
          }
        : null;
}
