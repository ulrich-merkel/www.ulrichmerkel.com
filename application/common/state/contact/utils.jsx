/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 *
 * @TODO Should be refactored and moved to actions
 *
 * @requires common/utils/validation
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';

/**
 * Validation helper to check all inputs.
 *
 * @function
 * @param {object} [state={}] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
function validate(state = {}) {
    return {
        name: required(state.name),
        email: required(state.email) && email(state.email),
        subject: required(state.subject),
        message: required(state.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @function
 * @param {object} state - The current redux contact state
 * @returns {boolean} The validation status
 */
function isValid(state) {
    const validated = validate(state);

    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

function canSendForm(state) {
    const isStateValid = isValid(state);

    return isStateValid
        && state.pristine
        && !state.pending
        && isBrowser(state);
}

export {
    canSendForm,
    isValid,
    validate
};
