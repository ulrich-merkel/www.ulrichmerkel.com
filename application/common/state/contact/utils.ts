/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';
import { ContactFormType, ValidateType } from './types';

/**
 * Validation helper to check all inputs.
 *
 * @param {object} [contactForm] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
export function validate(contactForm: ContactFormType): ValidateType {
    return {
        name: required(contactForm?.name),
        email: required(contactForm?.email) && email(contactForm?.email),
        subject: required(contactForm?.subject),
        message: required(contactForm?.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} The validation status
 */
export function isValid(contactForm: ContactFormType): boolean {
    const validated = validate(contactForm);
console.log({validated, contactForm})
    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

/**
 * Check if the form can be send.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} Whether the submit button is enabled or not
 */
export function canSendForm(contactForm: ContactFormType): boolean {
    const isStateValid = isValid(contactForm);

    return (
        isStateValid &&
        contactForm.pristine &&
        !contactForm.pending &&
        isBrowser(contactForm)
    );
}
