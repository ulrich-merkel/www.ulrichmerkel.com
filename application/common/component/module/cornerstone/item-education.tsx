/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent } from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import { Headline } from '../../element/headline';
import { ListItem } from '../../element/list-item';
import { P } from '../../element/paragraph';
import { Meta } from '../../element/meta';
import { View } from '../../element/view';
import { isValidArray } from '../../../utils/array';

type Props = {
    cssModifier?: string;
    description?: string[];
    headline?: string;
    lead?: string;
    offset?: string;
    place?: Record<string, unknown>;
    timeEnd?: string;
    timeStart?: string;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleCornerstoneItemEducation: FunctionComponent<Props> = (
    props
) => {
    const {
        cssModifier,
        description,
        headline,
        lead,
        offset,
        place,
        timeEnd,
        timeStart
    } = props;

    const composedListItemClassName = classnames(
        cssModifier
            ? `m-cornerstone__item--${cssModifier}`
            : 'm-cornerstone__item',
        offset ? `has-offset--${offset}` : ''
    );

    return (
        <ListItem
            className={composedListItemClassName}
            itemProp="itemListElement"
            itemType="https://schema.org/EducationEvent"
        >
            <View className="m-cornerstone__description">
                <View className="m-cornerstone__description-content">
                    <Headline
                        className="m-cornerstone__headline"
                        itemProp="name"
                        htmlElement="h4"
                    >
                        {headline}
                    </Headline>
                    <P
                        className="m-cornerstone__company"
                        itemProp="alternateName"
                    >
                        <strong>{lead}</strong>
                    </P>
                    <P className="m-cornerstone__time">
                        (
                        <time className="c-time" itemProp="startDate">
                            {timeStart}
                        </time>{' '}
                        -{' '}
                        <time className="c-time" itemProp="endDate">
                            {timeEnd}
                        </time>
                        )
                    </P>
                    {isValidArray(description) &&
                        description.map(function fnMap(text) {
                            /**
                             * DangerouslySetInnerHTML due to reacts double escaping. Otherwise html elements
                             * are not possible to be set via translation strings.
                             *
                             * @see {@link https://facebook.github.io/react/tips/dangerously-set-inner-html.html}
                             */
                            return (
                                <P
                                    key={shortid.generate()}
                                    className="m-cornerstone__text"
                                    itemProp="description"
                                    dangerouslySetInnerHTML={{ __html: text }}
                                />
                            );
                        })}
                </View>
            </View>
            <View className="m-cornerstone__bubble" />
            <View
                hidden
                itemProp="location"
                itemType="https://schema.org/Place"
            >
                <Meta itemProp="name" content={place.name} />
                <View
                    itemProp="address"
                    itemType="https://schema.org/PostalAddress"
                >
                    <Meta
                        itemProp="streetAddress"
                        content={place.streetAddress}
                    />
                    <Meta
                        itemProp="addressLocality"
                        content={place.addressLocality}
                    />
                    <Meta itemProp="sameAs" content={place.sameAs} />
                </View>
            </View>
        </ListItem>
    );
};
