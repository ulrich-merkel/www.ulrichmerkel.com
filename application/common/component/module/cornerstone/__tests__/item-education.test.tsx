import * as React from 'react';
import { render } from '../../../../../__tests__/utils/test-utils';
import { ModuleCornerstoneItemEducation } from '../item-education';

describe('ModuleCornerstoneItemEducation', function fnDescribe() {
    const props = {
        headline: 'Headline item education',
        lead: 'Lead  item education',
        timeStart: '20161212',
        timeEnd: '20161224',
        description: [
            'Description item education 1',
            'Description item education 2'
        ],
        place: {}
    };

    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <ModuleCornerstoneItemEducation {...props} />
        );
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render cssModifier and offset if passed', function fnIt() {
        const { asFragment } = render(
            <ModuleCornerstoneItemEducation
                {...props}
                cssModifier="education"
                offset="100"
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
