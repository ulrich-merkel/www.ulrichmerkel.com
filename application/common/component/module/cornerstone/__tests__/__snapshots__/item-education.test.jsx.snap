// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleCornerstoneItemEducation should render correctly 1`] = `
<DocumentFragment>
  <li
    class="m-cornerstone__item"
    itemprop="itemListElement"
    itemscope=""
    itemtype="https://schema.org/EducationEvent"
  >
    <div
      class="m-cornerstone__description"
    >
      <div
        class="m-cornerstone__description-content"
      >
        <h4
          class="is-centered c-type--h4 m-cornerstone__headline"
          itemprop="name"
        >
          Headline item education
        </h4>
        <p
          class="c-type--p m-cornerstone__company"
          itemprop="alternateName"
        >
          <strong>
            Lead  item education
          </strong>
        </p>
        <p
          class="c-type--p m-cornerstone__time"
        >
          (
          <time
            class="c-time"
            itemprop="startDate"
          >
            20161212
          </time>
           - 
          <time
            class="c-time"
            itemprop="endDate"
          >
            20161224
          </time>
          )
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item education 1
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item education 2
        </p>
      </div>
    </div>
    <div
      class="m-cornerstone__bubble"
    />
    <div
      hidden=""
      itemprop="location"
      itemscope=""
      itemtype="https://schema.org/Place"
    >
      <meta
        itemprop="name"
      />
      <div
        itemprop="address"
        itemscope=""
        itemtype="http://schema.org/PostalAddress"
      >
        <meta
          itemprop="streetAddress"
        />
        <meta
          itemprop="addressLocality"
        />
        <meta
          itemprop="sameAs"
        />
      </div>
    </div>
  </li>
</DocumentFragment>
`;

exports[`ModuleCornerstoneItemEducation should render cssModifier and offset if passed 1`] = `
<DocumentFragment>
  <li
    class="m-cornerstone__item--education has-offset--100"
    itemprop="itemListElement"
    itemscope=""
    itemtype="https://schema.org/EducationEvent"
  >
    <div
      class="m-cornerstone__description"
    >
      <div
        class="m-cornerstone__description-content"
      >
        <h4
          class="is-centered c-type--h4 m-cornerstone__headline"
          itemprop="name"
        >
          Headline item education
        </h4>
        <p
          class="c-type--p m-cornerstone__company"
          itemprop="alternateName"
        >
          <strong>
            Lead  item education
          </strong>
        </p>
        <p
          class="c-type--p m-cornerstone__time"
        >
          (
          <time
            class="c-time"
            itemprop="startDate"
          >
            20161212
          </time>
           - 
          <time
            class="c-time"
            itemprop="endDate"
          >
            20161224
          </time>
          )
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item education 1
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item education 2
        </p>
      </div>
    </div>
    <div
      class="m-cornerstone__bubble"
    />
    <div
      hidden=""
      itemprop="location"
      itemscope=""
      itemtype="https://schema.org/Place"
    >
      <meta
        itemprop="name"
      />
      <div
        itemprop="address"
        itemscope=""
        itemtype="http://schema.org/PostalAddress"
      >
        <meta
          itemprop="streetAddress"
        />
        <meta
          itemprop="addressLocality"
        />
        <meta
          itemprop="sameAs"
        />
      </div>
    </div>
  </li>
</DocumentFragment>
`;
